src/NetTopologySuite.IO.ShapeFile/Streams/ExternallyManagedStreamProvider.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+
+namespace NetTopologySuite.IO.Streams
+{
+    /// <summary>
+    /// A stream provider that provides <see cref="Stream"/>s who are managed by the application.
+    /// </summary>
+    public class ExternallyManagedStreamProvider : IStreamProvider
+    {
+        /// <summary>
+        /// Creates an instance of this class
+        /// </summary>
+        /// <param name="kind">The kind of stream</param>
+        /// <param name="stream">A <see cref="Stream"/> managed by the application</param>
+        public ExternallyManagedStreamProvider(string kind, Stream stream)
+        {
+            Kind = kind;
+            Stream = stream;
+        }
+
+        /// <summary>
+        /// Gets a value indicating that the underlying stream is read-only
+        /// </summary>
+        public bool UnderlyingStreamIsReadonly => !Stream.CanWrite;
+
+        protected Stream Stream { get; private set; }
+
+        /// <summary>
+        /// Function to return a Stream of the bytes
+        /// </summary>
+        /// <returns>An opened stream</returns>
+        public Stream OpenRead()
+        {
+            return Stream;
+        }
+
+        /// <summary>
+        /// Function to open the underlying stream for writing purposes
+        /// </summary>
+        /// <remarks>If <see cref="UnderlyingStreamIsReadonly"/> is not <value>true</value>
+        /// this method shall fail</remarks>
+        /// <returns>An opened stream</returns>
+        /// <exception cref="InvalidOperationException">Thrown if <see cref="UnderlyingStreamIsReadonly"/> is <value>true</value></exception>
+        public Stream OpenWrite(bool truncate)
+        {
+            if (UnderlyingStreamIsReadonly)
+                throw new InvalidOperationException();
+
+            if (truncate) {
+                Stream.SetLength(0);
```

Comment:
need to check `Stream.CanSeek` too

Reply:
In which cases? Do I need to check it like we check canWrite? or only when we want to truncate?

Reply:
only when we want to truncate.

Reply:
Ok, added + provided a test